cpp/src/io/shp/polygon_shapefile_reader.cpp

```diff
@@ -14,28 +14,48 @@
  * limitations under the License.
  */
 
+#include <cuspatial/detail/iterator.hpp>
 #include <cuspatial/error.hpp>
+#include <cuspatial/shapefile_reader.hpp>
 
 #include <ogrsf_frmts.h>
 
 #include <cudf/types.hpp>
 
+#include <thrust/iterator/counting_iterator.h>
+
 #include <memory>
 #include <string>
 #include <tuple>
 
 namespace {
 
+// TODO: a better approach is to precompute the total size needed by xs and ys
+// as a preprocess and preallocate xs, ys. `read_ring` should simply accept
+// an iterator range and return the pass-the-end iterator.
 cudf::size_type read_ring(OGRLinearRing const& ring,
                           std::vector<double>& xs,
-                          std::vector<double>& ys)
+                          std::vector<double>& ys,
+                          const cuspatial::winding_order ring_order)
 {
-  cudf::size_type num_vertices = ring.getNumPoints();
-
-  // append points in reverse order
-  for (cudf::size_type i = num_vertices - 1; i >= 0; i--) {
-    xs.push_back(ring.getX(i));
-    ys.push_back(ring.getY(i));
+  auto num_vertices      = ring.getNumPoints();
+  auto prev_num_vertices = xs.size();
+  xs.resize(xs.size() + num_vertices);
+  ys.resize(ys.size() + num_vertices);
+
+  auto output_it = thrust::make_zip_iterator(
+    thrust::make_tuple(xs.begin() + prev_num_vertices, ys.begin() + prev_num_vertices));
+  if (ring_order == cuspatial::winding_order::COUNTER_CLOCKWISE) {
+    auto ring_it = cuspatial::detail::make_counting_transform_iterator(
+      0, [&ring](auto i) { return thrust::make_tuple(ring.getX(i), ring.getY(i)); });
+
+    thrust::copy_n(ring_it, num_vertices, output_it);
```

Comment:
Both input and output data are on the host, right? If so, I would just use `std::copy_n` instead.